### ISSUE TITLE
SWMR has some problems:

### ISSUE THREAD

**Issue by NAThompson** [h5py/h5py#616]
Not unexpected as it's experimental, but:

1) If we use `hdf5-1.9.222-swmr0.tar.gz` ftp'd from ftp://ftp.hdfgroup.uiuc.edu/pub/outgoing/SWMR/src (which is linked from <a href="http://docs.h5py.org/en/latest/swmr.html">the documentation</a>) and the following dockerfile:

```<bash>
FROM ubuntu:15.04
ENV HOME /root
RUN apt-get update
RUN apt-get -yq install gcc build-essential subversion zlib1g-dev
ADD hdf5-1.9.222-swmr0.tar.gz /root
RUN cd ; cd hdf5-1.9.222-swmr0; ./configure --prefix=/usr/
RUN cd ; cd hdf5-1.9.222-swmr0; make && make install

RUN apt-get install -y python3-dev
RUN apt-get install -y python3-pip
RUN pip3 install Cython
RUN pip3 install numpy
RUN pip3 install six
RUN pip3 install h5py
CMD python3 -c "import h5py"
```

we get the error:

```<bash>
docker build -t hdf5v19 .
docker run hdf5v19
ValueError: Not property objects (Not property objects)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/h5py/__init__.py", line 31, in <module>
    from .highlevel import *
  File "/usr/local/lib/python3.4/dist-packages/h5py/highlevel.py", line 13, in <module>
    from ._hl.base import is_hdf5, HLObject
  File "/usr/local/lib/python3.4/dist-packages/h5py/_hl/base.py", line 78, in <module>
    dlapl = default_lapl()
  File "/usr/local/lib/python3.4/dist-packages/h5py/_hl/base.py", line 65, in default_lapl
    lapl = h5p.create(h5p.LINK_ACCESS)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/tmp/pip-build-wf8df8jq/h5py/h5py/_objects.c:2574)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/tmp/pip-build-wf8df8jq/h5py/h5py/_objects.c:2533)
  File "h5py/h5p.pyx", line 131, in h5py.h5p.create (/tmp/pip-build-wf8df8jq/h5py/h5py/h5p.c:2607)
  File "h5py/h5p.pyx", line 72, in h5py.h5p.propwrap (/tmp/pip-build-wf8df8jq/h5py/h5py/h5p.c:2407)
ValueError: Not a property list class (Not a property list class)
```
2) If we use the revision linked in <a href="http://docs.h5py.org/en/latest/swmr.html">the documentation</a>, namely, 27721, then we replace the Dockerfile becomes

```<bash>
FROM ubuntu:15.04
ENV HOME /root
RUN apt-get update
RUN apt-get -yq install gcc build-essential subversion zlib1g-dev
RUN cd ; svn co -r27721 http://svn.hdfgroup.uiuc.edu/hdf5/features/vol/ hdf5-vol
RUN cd ; cd hdf5-vol; ./configure --prefix=/usr/
RUN cd ; cd hdf5-vol; make -j8 && make install

RUN apt-get install -y python3-dev
RUN apt-get install -y python3-pip
RUN pip3 install Cython
RUN pip3 install numpy
RUN pip3 install six
RUN pip3 install h5py
```

However, the pip3 build breaks with the following error:

```<bash>
/tmp/pip-build-0xpasado/h5py/h5py/h5f.c: In function 'PyInit_h5f':

/tmp/pip-build-0xpasado/h5py/h5py/h5f.c:5380:36: error: 'H5F_ACC_SWMR_WRITE' undeclared (first use in this function)

   __pyx_t_8 = __Pyx_PyInt_From_int(H5F_ACC_SWMR_WRITE); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 37; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

                                    ^

/tmp/pip-build-0xpasado/h5py/h5py/h5f.c:5380:36: note: each undeclared identifier is reported only once for each function it appears in

/tmp/pip-build-0xpasado/h5py/h5py/h5f.c:5392:36: error: 'H5F_ACC_SWMR_READ' undeclared (first use in this function)

   __pyx_t_8 = __Pyx_PyInt_From_int(H5F_ACC_SWMR_READ); if (unlikely(!__pyx_t_8)) {__pyx_filename = __pyx_f[0]; __pyx_lineno = 38; __pyx_clineno = __LINE__; goto __pyx_L1_error;}

                                    ^

In file included from /usr/local/lib/python3.4/dist-packages/numpy/core/include/numpy/ndarrayobject.h:26:0,

                 from /usr/local/lib/python3.4/dist-packages/numpy/core/include/numpy/arrayobject.h:4,

                 from /tmp/pip-build-0xpasado/h5py/h5py/api_compat.h:26,

                 from /tmp/pip-build-0xpasado/h5py/h5py/h5f.c:280:
```

Looking the the source code for HDF5, it looks as if the SWMR functionality is not present in this revision, so the error is unsurprising. 

Is there an SVN revision you are using to test this feature?

(Really appreciate `h5py` BTW; incredibly well designed.)

**Comment by epourmal** [comment]
Hmm.. Are you sure you are using the correct HDF5 distribution? 

I see the symbol in the public header file

elenas-air:src epourmal$ pwd
/Users/epourmal/Downloads/hdf5-1.9.222-swmr0/src
elenas-air:src epourmal$ !grep
grep H5F_ACC_SWMR_WRITE *.h
H5Fpublic.h:#define H5F_ACC_SWMR_WRITE	(H5CHECK 0x0020u) /*indicate that this file is

May be you need a special version of h5py?

**Comment by gpwright** [comment]
I think if you pip install h5py like that then you won't be on the 2.5 branch.  You need to checkout version 2.5 of h5py from git and install it against your custom built SWMR build of hdf5.

I have been using this on my own projects and the 1.9.178 build of HDF5 combined with h5py 2.5 works fine except for one issue where you run out of resources after creating a large number of objects.  However the 1.9.222 build of HDF5 although fixing the resource issue is not currently compatible with h5py.  See https://github.com/h5py/h5py/issues/599

**Comment by NAThompson** [comment]
Ok, so at the advice of @gpwright I have changed my Dockerfile to:

```<bash>
FROM ubuntu:15.04
ENV HOME /root
RUN apt-get update
RUN apt-get -yq install gcc build-essential subversion zlib1g-dev

RUN apt-get install -y python3-dev
RUN apt-get install -y python3-pip
RUN apt-get install -y git
RUN pip3 install Cython
RUN pip3 install numpy
RUN pip3 install six

ADD hdf5-1.9.222-swmr0.tar.gz /root
RUN cd ; cd hdf5-1.9.222-swmr0; ./configure --prefix=/usr/
RUN cd ; cd hdf5-1.9.222-swmr0; make -j8 && make install


RUN git clone https://github.com/h5py/h5py.git /root/h5py
RUN cd /root/h5py; git checkout 2.5.0rc1
RUN cd /root/h5py; python3 setup.py configure --hdf5=/root/hdf5-1.9.222-swmr0
RUN cd /root/h5py; python3 setup.py build
RUN cd /root/h5py; python3 setup.py install

CMD python3 -c "import h5py"
```
Same error message:

```<bash>
# docker build -t hdf5v19 .
# docker run hdf5v19
ValueError: Not property objects (Not property objects)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python3.4/dist-packages/h5py-2.5.0-py3.4-linux-x86_64.egg/h5py/__init__.py", line 31, in <module>
    from .highlevel import *
  File "/usr/local/lib/python3.4/dist-packages/h5py-2.5.0-py3.4-linux-x86_64.egg/h5py/highlevel.py", line 13, in <module>
    from ._hl.base import is_hdf5, HLObject
  File "/usr/local/lib/python3.4/dist-packages/h5py-2.5.0-py3.4-linux-x86_64.egg/h5py/_hl/base.py", line 78, in <module>
    dlapl = default_lapl()
  File "/usr/local/lib/python3.4/dist-packages/h5py-2.5.0-py3.4-linux-x86_64.egg/h5py/_hl/base.py", line 65, in default_lapl
    lapl = h5p.create(h5p.LINK_ACCESS)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/root/h5py/h5py/_objects.c:2577)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/root/h5py/h5py/_objects.c:2536)
  File "h5py/h5p.pyx", line 131, in h5py.h5p.create (/root/h5py/h5py/h5p.c:2610)
  File "h5py/h5p.pyx", line 72, in h5py.h5p.propwrap (/root/h5py/h5py/h5p.c:2410)
ValueError: Not a property list class (Not a property list class)
```
You can change `git checkout 2.5.0rc1`` to ``git checkout 2.5.0`` and the same problem occurs.

**Comment by gpwright** [comment]
I think you may have better luck with version 1.9.178 of HDF5.  Version 1.9.222 is not currently compatible with h5py.  See the bug I filed #599 

I'm using SWMR with HDF5 1.9.178, and h5py version 2.5.  I'm also using python 2.7 rather than 3.4, but I don't think that would matter.

**Comment by gpwright** [comment]
Note: if you do switch to 1.9.178 it has an annoying bug that was fixed in 1.9.222 but chances are you won't encounter it as you have to open a large number of files or create a large number of objects.  It's the same bug that's described here: #480 it was fixed last year but the SWMR branch was forked before the fix.  They included the fix in 1.9.222, but made a bunch of other changes that unfortunately breaks h5py.

**Comment by NAThompson** [comment]
@gpwright: I can't find a link to 1.9.178; can I get it via ftp or is there a revision number?

**Comment by epourmal** [comment]
Please download it from here  ftp://ftp.hdfgroup.uiuc.edu/pub/outgoing/SWMR/src/OLD/

**Comment by gpwright** [comment]
@NAThompson I found time to look into this today and there is a simple fix.  Using h5py version 2.5 and HDF5 version 1.9.222 make one small code change to change the type of hid_t from int to long int.  Then build h5py and everything should be working.


diff --git a/h5py/api_types_hdf5.pxd b/h5py/api_types_hdf5.pxd
index a2174be..250d91e 100644
--- a/h5py/api_types_hdf5.pxd
+++ b/h5py/api_types_hdf5.pxd
@@ -11,7 +11,7 @@ from api_types_ext cimport *

 cdef extern from "hdf5.h":
   # Basic types
-  ctypedef long int hid_t
+  ctypedef int hid_t
   ctypedef int hbool_t
   ctypedef int herr_t
   ctypedef int htri_t

**Comment by andrewcollette** [comment]
Should be fixed by #625.
